File: backend/app/schemas/audit.py
"""Pydantic schemas for audit endpoints."""

from datetime import datetime
from typing import Literal, Any
from pydantic import BaseModel, Field


# =============================================================================
# Enums and Types
# =============================================================================

TrustLevelType = Literal["authoritative", "high", "medium", "low", "untrusted"]
DifferenceType = Literal["exact", "equivalent", "within_tolerance", "overlapping", "adjacent", "partial", "format_diff", "mismatch", "missing"]
DiscrepancyStatus = Literal["pending", "accepted", "rejected", "ignored"]
EnrichmentStatus = Literal["pending", "applied", "rejected", "ignored"]
AuditRunStatus = Literal["running", "completed", "failed", "cancelled"]
AuditScope = Literal["single", "selected", "all"]


# =============================================================================
# Discrepancy Schemas
# =============================================================================

class DiscrepancyBase(BaseModel):
    """Base schema for discrepancy records."""
    field_name: str
    current_value: str | None = None
    auction_value: str | None = None
    similarity: float | None = None
    difference_type: DifferenceType | None = None
    comparison_notes: str | None = None
    source_house: str
    trust_level: TrustLevelType
    auto_acceptable: bool = False


class DiscrepancyCreate(DiscrepancyBase):
    """Schema for creating discrepancy records."""
    coin_id: int
    auction_data_id: int | None = None
    audit_run_id: int | None = None


class DiscrepancyOut(DiscrepancyBase):
    """Schema for discrepancy output."""
    id: int
    coin_id: int
    auction_data_id: int | None = None
    audit_run_id: int | None = None
    normalized_current: str | None = None
    normalized_auction: str | None = None
    status: DiscrepancyStatus
    resolved_at: datetime | None = None
    resolution: str | None = None
    resolution_notes: str | None = None
    created_at: datetime
    
    # Extended data from related tables
    source_url: str | None = None
    auction_images: list[str] = []
    coin_primary_image: str | None = None
    
    # Coin details for preview
    coin_ruler: str | None = None
    coin_denomination: str | None = None
    coin_metal: str | None = None
    coin_grade: str | None = None
    coin_category: str | None = None
    coin_mint_year_start: int | None = None
    coin_mint_year_end: int | None = None
    coin_is_circa: bool | None = None
    
    class Config:
        from_attributes = True


class DiscrepancyDetail(DiscrepancyOut):
    """Detailed discrepancy with coin info."""
    coin_name: str | None = None
    auction_house: str | None = None
    auction_date: str | None = None


class DiscrepancyResolveRequest(BaseModel):
    """Request to resolve a discrepancy."""
    resolution: Literal["accepted", "rejected", "ignored"]
    notes: str | None = None


class BulkResolveRequest(BaseModel):
    """Request to bulk resolve discrepancies."""
    discrepancy_ids: list[int]
    resolution: Literal["accepted", "rejected", "ignored"]
    notes: str | None = None


class PaginatedDiscrepancies(BaseModel):
    """Paginated discrepancy response."""
    items: list[DiscrepancyOut]
    total: int
    page: int
    per_page: int
    pages: int


# =============================================================================
# Enrichment Schemas
# =============================================================================

class EnrichmentBase(BaseModel):
    """Base schema for enrichment records."""
    field_name: str
    suggested_value: str
    source_house: str
    trust_level: TrustLevelType
    confidence: float | None = None
    auto_applicable: bool = False


class EnrichmentCreate(EnrichmentBase):
    """Schema for creating enrichment records."""
    coin_id: int
    auction_data_id: int | None = None
    audit_run_id: int | None = None


class EnrichmentOut(EnrichmentBase):
    """Schema for enrichment output."""
    id: int
    coin_id: int
    auction_data_id: int | None = None
    audit_run_id: int | None = None
    status: EnrichmentStatus
    applied: bool
    applied_at: datetime | None = None
    rejection_reason: str | None = None
    created_at: datetime
    
    # Extended data from related tables
    source_url: str | None = None
    auction_images: list[str] = []
    coin_primary_image: str | None = None
    
    # Coin details for preview
    coin_ruler: str | None = None
    coin_denomination: str | None = None
    coin_metal: str | None = None
    coin_grade: str | None = None
    coin_category: str | None = None
    coin_mint_year_start: int | None = None
    coin_mint_year_end: int | None = None
    coin_is_circa: bool | None = None
    
    class Config:
        from_attributes = True


class EnrichmentDetail(EnrichmentOut):
    """Detailed enrichment with coin info."""
    coin_name: str | None = None
    auction_house: str | None = None


class EnrichmentApplyRequest(BaseModel):
    """Request to apply enrichments."""
    enrichment_ids: list[int]


class EnrichmentRejectRequest(BaseModel):
    """Request to reject enrichments."""
    enrichment_ids: list[int]
    reason: str | None = None


class PaginatedEnrichments(BaseModel):
    """Paginated enrichment response."""
    items: list[EnrichmentOut]
    total: int
    page: int
    per_page: int
    pages: int


# =============================================================================
# Audit Run Schemas
# =============================================================================

class AuditRunCreate(BaseModel):
    """Schema for starting an audit run."""
    scope: AuditScope
    coin_ids: list[int] | None = None


class AuditRunOut(BaseModel):
    """Schema for audit run output."""
    id: int
    scope: AuditScope
    coin_ids: list[int] | None = None
    status: AuditRunStatus
    
    # Progress
    total_coins: int
    coins_audited: int
    coins_with_issues: int
    
    # Results
    discrepancies_found: int
    enrichments_found: int
    images_downloaded: int
    auto_accepted: int
    auto_applied: int
    
    # Timing
    started_at: datetime
    completed_at: datetime | None = None
    error_message: str | None = None
    
    class Config:
        from_attributes = True


class AuditRunProgress(BaseModel):
    """Progress update for an audit run."""
    id: int
    status: AuditRunStatus
    progress_percent: float
    coins_audited: int
    total_coins: int
    discrepancies_found: int
    enrichments_found: int


# =============================================================================
# Audit Summary Schemas
# =============================================================================

class AuditSummary(BaseModel):
    """Summary statistics for audit dashboard."""
    pending_discrepancies: int
    pending_enrichments: int
    discrepancies_by_trust: dict[str, int]
    discrepancies_by_field: dict[str, int]
    discrepancies_by_source: dict[str, int]
    recent_runs: list[dict]


class CoinAuditSummary(BaseModel):
    """Audit summary for a single coin."""
    coin_id: int
    auctions_checked: int
    discrepancies: int
    enrichments: int
    message: str | None = None


# =============================================================================
# Image Download Schemas
# =============================================================================

class ImageDownloadRequest(BaseModel):
    """Request to download auction images."""
    auction_data_id: int | None = None  # If None, download from all auctions


class ImageDownloadResult(BaseModel):
    """Result of image download operation."""
    coin_id: int
    auctions_processed: int
    images_downloaded: int
    duplicates_skipped: int
    errors: int


# =============================================================================
# Filter Schemas
# =============================================================================

class DiscrepancyFilters(BaseModel):
    """Filters for discrepancy queries."""
    status: DiscrepancyStatus | None = None
    field_name: str | None = None
    source_house: str | None = None
    trust_level: TrustLevelType | None = None
    coin_id: int | None = None
    audit_run_id: int | None = None
    min_similarity: float | None = None
    max_similarity: float | None = None


class EnrichmentFilters(BaseModel):
    """Filters for enrichment queries."""
    status: EnrichmentStatus | None = None
    field_name: str | None = None
    source_house: str | None = None
    trust_level: TrustLevelType | None = None
    coin_id: int | None = None
    audit_run_id: int | None = None
    auto_applicable: bool | None = None

File: frontend/src/types/audit.ts
/**
 * TypeScript types for audit functionality.
 */

export type TrustLevel = "authoritative" | "high" | "medium" | "low" | "untrusted";
export type DifferenceType = 
  | "exact" 
  | "equivalent" 
  | "within_tolerance" 
  | "overlapping" 
  | "adjacent" 
  | "partial" 
  | "format_diff" 
  | "mismatch" 
  | "missing";
export type DiscrepancyStatus = "pending" | "accepted" | "rejected" | "ignored";
export type EnrichmentStatus = "pending" | "applied" | "rejected" | "ignored";
export type AuditRunStatus = "running" | "completed" | "failed" | "cancelled";
export type AuditScope = "single" | "selected" | "all";

/**
 * Discrepancy record - data conflict between coin and auction
 */
export interface Discrepancy {
  id: number;
  coin_id: number;
  auction_data_id: number | null;
  audit_run_id: number | null;
  field_name: string;
  current_value: string | null;
  auction_value: string | null;
  normalized_current: string | null;
  normalized_auction: string | null;
  similarity: number | null;
  difference_type: DifferenceType | null;
  comparison_notes: string | null;
  source_house: string;
  trust_level: TrustLevel;
  auto_acceptable: boolean;
  status: DiscrepancyStatus;
  resolved_at: string | null;
  resolution: string | null;
  resolution_notes: string | null;
  created_at: string;
  // Extended data from related tables
  source_url?: string | null;
  auction_images?: string[];
  coin_primary_image?: string | null;
  // Coin details
  coin_ruler?: string | null;
  coin_denomination?: string | null;
  coin_metal?: string | null;
  coin_grade?: string | null;
  coin_category?: string | null;
  coin_mint_year_start?: number | null;
  coin_mint_year_end?: number | null;
  coin_is_circa?: boolean;
}

/**
 * Enrichment record - suggested field update from auction data
 */
export interface Enrichment {
  id: number;
  coin_id: number;
  auction_data_id: number | null;
  audit_run_id: number | null;
  field_name: string;
  suggested_value: string;
  source_house: string;
  trust_level: TrustLevel;
  confidence: number | null;
  auto_applicable: boolean;
  status: EnrichmentStatus;
  applied: boolean;
  applied_at: string | null;
  rejection_reason: string | null;
  created_at: string;
  // Extended data from related tables
  source_url?: string | null;
  auction_images?: string[];
  coin_primary_image?: string | null;
  // Coin details
  coin_ruler?: string | null;
  coin_denomination?: string | null;
  coin_metal?: string | null;
  coin_grade?: string | null;
  coin_category?: string | null;
  coin_mint_year_start?: number | null;
  coin_mint_year_end?: number | null;
  coin_is_circa?: boolean;
}

/**
 * Audit run - tracks an audit execution session
 */
export interface AuditRun {
  id: number;
  scope: AuditScope;
  coin_ids: number[] | null;
  status: AuditRunStatus;
  total_coins: number;
  coins_audited: number;
  coins_with_issues: number;
  discrepancies_found: number;
  enrichments_found: number;
  images_downloaded: number;
  auto_accepted: number;
  auto_applied: number;
  started_at: string;
  completed_at: string | null;
  error_message: string | null;
}

/**
 * Audit run progress for polling
 */
export interface AuditRunProgress {
  id: number;
  status: AuditRunStatus;
  progress_percent: number;
  coins_audited: number;
  total_coins: number;
  discrepancies_found: number;
  enrichments_found: number;
}

/**
 * Audit summary statistics
 */
export interface AuditSummary {
  pending_discrepancies: number;
  pending_enrichments: number;
  discrepancies_by_trust: Record<string, number>;
  discrepancies_by_field: Record<string, number>;
  discrepancies_by_source: Record<string, number>;
  recent_runs: Array<{
    id: number;
    scope: string;
    status: string;
    started_at: string | null;
    coins_audited: number;
    discrepancies_found: number;
  }>;
}

/**
 * Coin audit summary
 */
export interface CoinAuditSummary {
  coin_id: number;
  auctions_checked: number;
  discrepancies: number;
  enrichments: number;
  message?: string;
}

/**
 * Paginated response wrapper
 */
export interface PaginatedResponse<T> {
  items: T[];
  total: number;
  page: number;
  per_page: number;
  pages: number;
}

/**
 * Filter options for discrepancies
 */
export interface DiscrepancyFilters {
  status?: DiscrepancyStatus;
  field_name?: string;
  source_house?: string;
  trust_level?: TrustLevel;
  coin_id?: number;
  audit_run_id?: number;
  min_similarity?: number;
  max_similarity?: number;
  page?: number;
  per_page?: number;
  sort_by?: string;
  sort_order?: "asc" | "desc";
}

/**
 * Filter options for enrichments
 */
export interface EnrichmentFilters {
  status?: EnrichmentStatus;
  field_name?: string;
  source_house?: string;
  trust_level?: TrustLevel;
  coin_id?: number;
  audit_run_id?: number;
  auto_applicable?: boolean;
  page?: number;
  per_page?: number;
  sort_by?: string;
  sort_order?: "asc" | "desc";
}

/**
 * Request to start an audit
 */
export interface StartAuditRequest {
  scope: AuditScope;
  coin_ids?: number[];
}

/**
 * Request to resolve a discrepancy
 */
export interface ResolveDiscrepancyRequest {
  resolution: "accepted" | "rejected" | "ignored";
  notes?: string;
}

/**
 * Image download result
 */
export interface ImageDownloadResult {
  coin_id: number;
  auctions_processed: number;
  images_downloaded: number;
  duplicates_skipped: number;
  errors: number;
}

/**
 * Trust level badge color mapping - works in both light and dark mode
 * Uses outline style with colored text for better visibility
 */
export const TRUST_LEVEL_COLORS: Record<TrustLevel, string> = {
  authoritative: "bg-amber-500/15 text-amber-700 dark:text-amber-400 border-amber-500/40",
  high: "bg-emerald-500/15 text-emerald-700 dark:text-emerald-400 border-emerald-500/40",
  medium: "bg-sky-500/15 text-sky-700 dark:text-sky-400 border-sky-500/40",
  low: "bg-orange-500/15 text-orange-700 dark:text-orange-400 border-orange-500/40",
  untrusted: "bg-slate-500/15 text-slate-700 dark:text-slate-400 border-slate-500/40",
};

/**
 * Difference type display labels
 */
export const DIFFERENCE_TYPE_LABELS: Record<DifferenceType, string> = {
  exact: "Exact Match",
  equivalent: "Equivalent",
  within_tolerance: "Within Tolerance",
  overlapping: "Overlapping",
  adjacent: "Adjacent",
  partial: "Partial Match",
  format_diff: "Format Difference",
  mismatch: "Mismatch",
  missing: "Missing",
};

/**
 * LLM-suggested rarity information
 */
export interface LLMRarityInfo {
  rarity_code: string | null;
  rarity_description: string | null;
  specimens_known: number | null;
  source: string | null;
}

/**
 * Catalog reference validation result
 */
export interface CatalogReferenceValidation {
  reference_text: string;
  parsed_catalog: string | null;
  parsed_number: string | null;
  parsed_volume: string | null;
  validation_status: "matches" | "partial_match" | "mismatch" | "unknown";
  confidence: number;
  match_reason: string | null;
  existing_reference: string | null;
  /** Category/catalog consistency warning from backend catalog_validation */
  numismatic_warning?: string | null;
}

/**
 * LLM-suggested design (legends, exergue, descriptions)
 */
export interface LlmSuggestedDesign {
  obverse_legend?: string | null;
  reverse_legend?: string | null;
  exergue?: string | null;
  obverse_description?: string | null;
  reverse_description?: string | null;
  obverse_legend_expanded?: string | null;
  reverse_legend_expanded?: string | null;
}

/**
 * LLM-suggested attribution (issuer, mint, denomination, dates)
 */
export interface LlmSuggestedAttribution {
  issuer?: string | null;
  mint?: string | null;
  denomination?: string | null;
  year_start?: number | null;
  year_end?: number | null;
}

/**
 * LLM suggestion item for review queue
 */
export interface LLMSuggestionItem {
  coin_id: number;
  // Core coin identification
  issuer: string | null;
  denomination: string | null;
  mint: string | null;
  year_start: number | null;
  year_end: number | null;
  category: string | null;
  // Legends for context
  obverse_legend: string | null;
  reverse_legend: string | null;
  // Existing catalog references (for comparison)
  existing_references: string[];
  // LLM suggestions
  suggested_references: string[];
  validated_references: CatalogReferenceValidation[];
  rarity_info: LLMRarityInfo | null;
  /** Design suggestions (legends, exergue, descriptions) from transcribe/identify */
  suggested_design?: LlmSuggestedDesign | null;
  /** Attribution suggestions (issuer, mint, denomination, dates) from identify */
  suggested_attribution?: LlmSuggestedAttribution | null;
  enriched_at: string | null;
}

/**
 * LLM review queue response
 */
export interface LLMReviewQueueResponse {
  items: LLMSuggestionItem[];
  total: number;
}
